File: go/internal/cli/reindex.go
package cli

import (
	"fmt"

	"github.com/spf13/cobra"
)

// newReindexCmd rebuilds the derived SQLite sidecar index.
//
// Needed when enabling `index = "sqlite"` on an existing store (nothing has
// been indexed yet), and the only way back after the index has been disabled —
// which happens whenever it disagrees with the backend, so that a stale index
// makes reads slower rather than wrong.
func newReindexCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "reindex",
		Short: "Rebuild the sidecar index from the vector store",
		RunE: func(cmd *cobra.Command, _ []string) error {
			store := storeFromCtx(cmd.Context())
			if store.Index() == nil {
				// The root command builds the store from config, so an
				// unconfigured index means there is nothing to rebuild — but
				// the user clearly wants one, so build it here rather than
				// making them edit config first.
				if err := store.EnableIndex(cmd.Context(), ""); err != nil {
					return err
				}
			}
			res, err := store.Reindex(cmd.Context())
			if err != nil {
				return err
			}
			if fmtFromCtx(cmd.Context()) == FormatJSON {
				return printJSON(res)
			}
			if !res.Enabled {
				return fmt.Errorf("%s", res.Error)
			}
			fmt.Printf("Indexed %d memories → %s\n", res.Indexed, res.Path)
			fts := "no (SQLite built without FTS5)"
			if res.FTS {
				fts = "yes"
			}
			fmt.Printf("  full-text search: %s\n", fts)
			if !res.Healthy {
				return fmt.Errorf("index still unhealthy — %s", res.Error)
			}
			return nil
		},
	}
}

File: ai_houkai/cli/commands/reindex.py
"""Rebuild the derived SQLite sidecar index."""

from __future__ import annotations

import json as jsonlib

import typer

from ai_houkai.memory_system import MemoryStore


def reindex(
    ctx: typer.Context,
    json: bool = typer.Option(False, "--json", help="Emit raw JSON"),
) -> None:
    """Rebuild the sidecar index from the Chroma collection.

    Needed when enabling `index = "sqlite"` on an existing store (nothing has
    been indexed yet), and the only way back after the index has been disabled
    — which happens whenever it disagrees with Chroma, so that a stale index
    makes reads slower rather than wrong.
    """
    cfg = ctx.obj["config"]
    store: MemoryStore = ctx.obj["store"]

    if store.index is None:
        # The CLI callback builds the store from config, so an unconfigured
        # index means there is nothing to rebuild — but the user clearly wants
        # one, so build it here rather than making them edit config first.
        store = MemoryStore(
            path=cfg.store_path, collection=cfg.collection, actor="cli",
            index="sqlite",
        )
        ctx.obj["store"] = store

    result = store.reindex()
    if json:
        typer.echo(jsonlib.dumps(result, indent=2))
    else:
        if not result.get("enabled"):
            typer.echo(f"Error: {result.get('error')}", err=True)
            raise typer.Exit(1)
        typer.echo(f"Indexed {result['indexed']} memories → {result['path']}")
        typer.echo(f"  full-text search: {'yes' if result['fts'] else 'no (SQLite built without FTS5)'}")
        if not result["healthy"]:
            typer.echo(f"  WARNING: index still unhealthy — {result['error']}",
                       err=True)
    if not result.get("healthy"):
        raise typer.Exit(1)
